File: Python/Assignment2/Shapes3D.py
import math
from Shapes import Polygon

class Shape:
    def __init__(self):
        raise NotImplementedError()
    
    def getSurfaceArea(self):
        pass

    def getVolume(self):
        pass

class Cuboid(Shape):
    def __init__(self, width, height, depth):
        self._width = width
        self._height = height
        self._depth = depth
    
    def getSurfaceArea(self):
        return 2 * (self._height * self._width + self._depth * self._height + self._width * self._depth)
    
    def getVolume(self):
        return self._height * self._width * self._depth
    
class Cube(Cuboid):
    def __init__(self, sidelength):
        super().__init__(sidelength, sidelength, sidelength)

class Cylinder(Shape):
    def __init__(self, radius, height):
        self._radius = radius
        self._height = height

    def getSurfaceArea(self):
        return (2 * math.pi * self._radius * self._height) + (2 * math.pi * (self._radius**2))
    
    def getVolume(self):
        return math.pi * (self._radius ** 2) * self._height

class Sphere(Shape):
    def __init__(self, radius):
        self._radius = radius

    def getSurfaceArea(self):
        return 4 * math.pi * (self._radius ** 2)
    
    def getVolume(self):
        return (4/3) * math.pi * (self._radius ** 3)

class Prism(Shape):
    def __init__(self, sidelength, faces, height):
        self._sidelength = sidelength
        self._faces = faces
        self._height = height
        self._base = Polygon(sideLength = sidelength,sides = faces)
        self._area = (2 * self._base.GetArea()) + (self._base.GetPerimeter() * self._height)
        self._volume = self._base.GetArea() * self._height

    def getSurfaceArea(self):
        return self._area
    
    def getVolume(self):
        return self._volume

File: CSharp/ParsingShapes/Shapes3D.cs
namespace David.Shapes3D {

    public abstract class Shape3D {
        public abstract double GetSurfaceArea();
        public abstract double GetVolume();
    }

    class Cuboid : Shape3D {
        public double width;
        public double height;
        public double depth;

        public Cuboid(double width, double height, double depth) {
            this.width = width;
            this.height = height;
            this.depth = depth;
        }
        override public double GetSurfaceArea() {
            return 2 * (this.width * this.height + this.height * this.depth + this.width * this.depth);
        }
        override public double GetVolume() {
            return this.width * this.height * this.depth;
        }
    }

    class Cube : Cuboid {
        public Cube(double length) : base(length, length, length) {}
    }

    class Cylinder : Shape3D {
        public double radius;
        public double height;

        public Cylinder(double radius, double height) {
            this.radius = radius;
            this.height = height;
        }
        override public double GetSurfaceArea() {
            return (2 *  Math.PI * this.radius * this.height) + (2 * Math.PI * Math.Pow(this.radius, 2));
        }
        override public double GetVolume() {
            return Math.PI * Math.Pow(this.radius, 2) * this.height;
        }
    }

    class Sphere : Shape3D {
        public double radius;

        public Sphere(double radius) {
            this.radius = radius;
        }
        public override double GetSurfaceArea() {
            return 4 * Math.PI * Math.Pow(this.radius, 2);
        }
        public override double GetVolume() {
            return (4d / 3d) * Math.PI * Math.Pow(this.radius, 3);
        }
    }
}